tasks/deploy/github.py

```diff
@@ -0,0 +1,157 @@
+        return self.make_request(path, json=True)
+
+    def trigger_workflow(self, repo_name, workflow_name, ref, inputs=None):
+        """
+        Create a pipeline targeting a given reference of a project.
+        ref must be a branch or a tag.
+        """
+        if inputs is None:
+            inputs = {}
+
+        path = "/repos/{}/actions/workflows/{}/dispatches".format(repo_name, workflow_name)
+        data = json.dumps({"ref": ref, "inputs": inputs})
+        return self.make_request(path, data=data)
+
+    def workflow_run(self, repo_name, run_id):
+        """
+        Gets info on a specific workflow.
+        """
+        path = "/repos/{}/actions/runs/{}".format(repo_name, run_id)
+        return self.make_request(path, json=True)
+
+    def download_artifact(self, repo_name, artifact_id, destination_dir):
+        """
+        Downloads the artifact identified by artifact_id to destination_dir.
+        """
+        path = "/repos/{}/actions/artifacts/{}/zip".format(repo_name, artifact_id)
+        content = self.make_request(path, raw_content=True)
+
+        zip_target_path = os.path.join(destination_dir, "{}.zip".format(artifact_id))
+        with open(zip_target_path, "wb") as f:
+            f.write(content)
+        return zip_target_path
+
+    def workflow_run_artifacts(self, repo_name, run_id):
+        """
+        Gets list of artifacts for a workflow run.
+        """
+        path = "/repos/{}/actions/runs/{}/artifacts".format(repo_name, run_id)
+        return self.make_request(path, json=True)
+
+    def latest_workflow_run_for_ref(self, repo_name, workflow_name, ref):
+        """
+        Gets latest workflow run for a given reference
+        """
+        runs = self.workflow_runs(repo_name, workflow_name)
+        ref_runs = [run for run in runs["workflow_runs"] if run["head_branch"] == ref]
+
+        if len(ref_runs) == 0:
+            return None
+
+        return sorted(ref_runs, key=lambda run: run['created_at'], reverse=True)[0]
+
+    def workflow_runs(self, repo_name, workflow_name):
+        """
+        Gets all workflow runs for a workflow.
+        """
+        path = "/repos/{}/actions/workflows/{}/runs".format(repo_name, workflow_name)
+        return self.make_request(path, json=True)
+
+    def make_request(self, path, headers=None, data=None, json=False, raw_content=False):
+        """
+        Utility to make a request to the Gitlab API.
+        """
+        import requests
+
+        url = self.BASE_URL + path
+
+        headers = dict(headers or [])
+        headers["Authorization"] = "token {}".format(self.api_token)
+        headers["Accept"] = "application/vnd.github.v3+json"
+        for _ in range(5):  # Retry up to 5 times
+            try:
+                if data:
+                    r = requests.post(url, headers=headers, data=data)
+                else:
+                    r = requests.get(url, headers=headers)
+                if r.status_code < 400:  # Success
```

Comment:
Are redirects handled automatically by `requests`?

Reply:
Yes, by default `requests` handles redirects. You need to pass `allow_redirects=False` to disable this: https://requests.readthedocs.io/en/master/user/quickstart/#redirection-and-history